Coerce mod agnostic

joanrue opened the pull request:
Coerce() function can now handle cupy and dask arrays.

joanrue committed:
[solver] add conjugate gradient method

Added the Conjugate Gradient Method iterative solver based on the Sepand;;s Solver.py base class

joanrue committed:
[abc/solver] add base class initialisation in _Worker

_Worker was not initializing base class threading.Thread, this was raising the error 'Thread.__init__() was not called'

joanrue committed:
[solver/cg] Change np.einsum to module agnostic funcs (__mul__ & __add__)

joanrue committed:
[solver/cg] Change numpy code to module agnostic based on input data

joanrue committed:
[solver/cg] Fix module agnosticity and precision enforcement.

Tested with numpy cupy and dask arrays

joanrue committed:
[runtime/init] Add module agnosticity for coerce() func

SepandKashani commented on src/pycsou/abc/solver.py:
```diff
@@ -566,6 +566,7 @@ def _m_persist(self):
 
     class _Worker(threading.Thread):
         def __init__(self, solver: "Solver"):
+            super().__init__()
```
Not really needed, but ok.

SepandKashani replied:
ok, based on our discussion.

SepandKashani commented on src/pycsou/runtime/__init__.py:
```diff
@@ -138,13 +139,21 @@ def coerce(x):
         Fails if operation is impossible or unsafe. (I.e. casting complex-valued data.)
     """
     dtype = getPrecision().value
+    try:
```
I don't quite agree with all the steps here. Let's talk about this in the afternoon.

SepandKashani reviewed (commented):
Have yet to look at `cg.py`.

joanrue committed:
[runtime/init] FIX: remove 'casting' argument from coerce function to allow coercing of cupy arrays

Also fixes the previous commit in which the array module is queried unnecessarily

SepandKashani commented on src/pycsou/runtime/__init__.py:
```diff
@@ -7,6 +7,7 @@
 
 import numpy as np
 
+import pycsou.util as pycu
```
Not required anymore.

SepandKashani reviewed (changes requested):
Some minor things to fix. (See comments.)
For `cg.py`, I suggest removing it entirely from this pull request and adding it only to #15

SepandKashani commented:
@joanrue Are you currently working on this?
Only `runtime/__init__.py` should be modified in this pull request. (No `xp` inference needed, just drop the `casting=` parameter and it should be good to go.

joanrue committed:
Merge branch 'v2-dev' into coerce_mod_agnostic

joanrue commented:
@SepandKashani Sorry made a mistake before you explained me the stash (This should not happen again). I fixed the conflict and I think this branch is ready to be merged.

SepandKashani reviewed (approved):
Looks good.
